demo.py

```diff
@@ -13,212 +13,96 @@
-    X_train = normalizer.fit_transform(X_train)
-
-    # X_test = imputer.fit_transform(X_test)
-    X_test = scaler.fit_transform(X_test)
-    X_test = normalizer.fit_transform(X_test)
-
-
-    # ===============================================
-    print('\nAdaboost:')
-
-    clf = AdaBoostClassifier(n_estimators=NUM_WEAK_CLASSIFIERS)
+from qboost import qboost_lambda_sweep
+from datasets import make_blob_data, get_handwritten_digits_data
 
-    # scores = cross_val_score(clf, X, y, cv=5, scoring='accuracy')
-    clf.fit(X_train, y_train)
 
-    hypotheses_ada = clf.estimators_
-    # clf.estimator_weights_ = np.random.uniform(0,1,size=NUM_WEAK_CLASSIFIERS)
-    y_train_pred = clf.predict(X_train)
-    y_test_pred = clf.predict(X_test)
-
-    print_accuracy(y_train, y_train_pred, y_test, y_test_pred)
-
-
-    # ===============================================
-    print('\nDecision tree:')
-
-    clf2 = WeakClassifiers(n_estimators=NUM_WEAK_CLASSIFIERS, max_depth=TREE_DEPTH)
-    clf2.fit(X_train, y_train)
-
-    y_train_pred2 = clf2.predict(X_train)
-    y_test_pred2 = clf2.predict(X_test)
-
-    if verbose:
-        print('weights:\n', clf2.estimator_weights)
-
-    print_accuracy(y_train, y_train_pred2, y_test, y_test_pred2)
-
-
-    # ===============================================
-    print('\nQBoost:')
-
-    DW_PARAMS = {'num_reads': NUM_READS,
-                 'auto_scale': True,
-                 # "answer_mode": "histogram",
-                 'num_spin_reversal_transforms': 10,
-                 # 'annealing_time': 10,
-                 }
-
-    clf3 = QBoostClassifier(n_estimators=NUM_WEAK_CLASSIFIERS, max_depth=TREE_DEPTH)
-    clf3.fit(X_train, y_train, emb_sampler, lmd=lmd, **DW_PARAMS)
-
-    y_train_dw = clf3.predict(X_train)
-    y_test_dw = clf3.predict(X_test)
-
-    if verbose:
-        print('weights\n', clf3.estimator_weights)
-
-    print_accuracy(y_train, y_train_dw, y_test, y_test_dw)
-
-
-    # ===============================================
-    print('\nQBoostPlus:')
-    clf4 = QBoostPlus([clf, clf2, clf3])
-    clf4.fit(X_train, y_train, emb_sampler, lmd=lmd, **DW_PARAMS)
-    y_train4 = clf4.predict(X_train)
-    y_test4 = clf4.predict(X_test)
+if __name__ == '__main__':
+    import argparse
 
-    if verbose:
-        print('weights\n', clf4.estimator_weights)
+    parser = argparse.ArgumentParser(description="Run QBoost example", epilog="Information about additional options that are specific to the data set can be obtained using either 'demo.py blobs -h' or 'demo.py digits -h'.")
+    parser.add_argument('--verbose', action='store_true')
 
-    print_accuracy(y_train, y_train4, y_test, y_test4)
+    subparsers = parser.add_subparsers(title='dataset', description='dataset to use', dest='dataset', required=True)
 
+    sp_blobs = subparsers.add_parser('blobs', help='blobs data set')
+    sp_blobs.add_argument('--num-samples', type=int, default=2000, help='number of samples (default: %(default)s)')
```

Comment:
"number of " --> "Number of 
Also below
Is it work using a smaller default value to get results faster or is 2000 close to the minimum for such training?

Reply:
The problem size scales with the number of features, not the number of samples.  As noted separately, I think the reason it appears to run slowly is just HSS overhead (which from my own testing, is quite variable; I just did a run that took 3m30s compared to prior runs taking around 2m20s).  Part of the issue is that each run of the demo actually solves the BQM 10 times as part of the parameter sweep.  If we wanted to speed it up, reducing the number of lambda values in the sweep would be the most effective (or adding an option to use simulated annealing...)

Reply:
Might be something to think about for all examples. The first examples in this repo were all QPU and ran fast. With HSS some examples can take minutes to produce results. For new users running an example in the IDE it might look as though it's hanging. 
Might be worth printing some sort of "Running... might take a few minutes" statement on such long-running problems.

Reply:
The first example is somewhat faster, depending on which case you run.  I ran the "mnist" case and it took about a minute, so it's not exactly super-fast ("wisc" ran faster).  Part of that may be using QPU directly instead of HSS, but another factor is that this update solves 10 BQM's per run, whereas previously I think it did 2 (QBoost and "QBoostPlust", each with fixed lambda value).

There are a couple of options to consider for speeding up the present code:
- Could switch from HSS to QPU solver.  May need to add in some extra checking, because the "blobs" data set has an option that allows the user to specify a value for number of features (variables).  If they use a large value, it might not fit on the QPU or embedding could take a long time.
- I could hard-code in a default lambda value for each problem.  This would get us a 10x speedup.  The sweep over lambda values could be moved under a command line option.

Reply:
Sorry for not being clearer, by "The first examples in this repo" I meant the first examples of the dwave-examples repo. Over the last year we've added HSS example that run slower. This example is not particularly slow compared to other HSS examples.
I was just wondering aloud whether for all such HSS examples in dwave-examples we should add some sort of "Running... might take a few minutes" printout when the demo is run that can take more than half a minute to print out some results.

Reply:
I agree the runtime of over 18 minutes is too much:
```
Leap IDE /workspace/qboost $ time python demo.py digits
Number of features: 64
Number of training samples: 216
Number of test samples: 144

Number of selected features: 39
Score on test set: 0.9861111111111112

real    18m25.049s
user    0m14.726s
sys     0m4.950s

Leap IDE /workspace/qboost $ time python demo.py blobs
Informative features: [0, 1]
Selected features: [0, 1]
Score on test set: 0.715

real    18m29.924s
user    0m14.313s
sys     0m5.024s
```

At the very least (if we can't shorten it), we should include a clear warning.

Reply:
This is my own ignorance, but do we expect a big discrepancy in runtimes for Leap IDE vs local?  Here is a run I just did that only took 2 minutes:

```
(skl-2.4) jmcfarland@debian:~/git/dwave-examples/qboost$ time python demo.py blobs
Informative features: [0, 1]
Selected features: [0, 1]
Score on test set: 0.68875

real    2m5.718s
user    0m5.735s
sys     0m1.261s
```

Reply:
LeapIDE is limited CPU/memory/network wise, probably much worse that your desktop, so if a demo is compute-intensive, than yes. The previous version (in master) runs relatively quickly comparing to this. Not sure we can do much, but if not, let's just include a warning message.

Reply:
OK, I went ahead and did both (speedup and message): I changed the behavior so that the default is to run only a single solution.  Locally, this runs in about 20s for me (should be about a 10x speedup).  I added a new command line option, `--cross-validation`, which provides the previous behavior of running 10 solutions to find the best value for lambda.

In addition, I added a message when `--cross-validation` is used:
```
Performing cross-validation using 10 values of lambda, this make take several minutes...
```